File: leetcode_problems/merge_Two_Sorted_Lists.cpp
#include <iostream>

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
 
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* lastNode;
        ListNode* temp = new ListNode();
        lastNode = temp;
        while (true) {
            if (list1 == NULL) {
                lastNode->next = list2;
                break;
            } else if (list2 == NULL) {
                lastNode->next = list1;
                break;
            }    
            if (list1->val <= list2->val) {
                lastNode->next = list1;
                list1 = list1->next;
            } else {
                lastNode->next = list2;
                list2 = list2->next;
            }

            lastNode = lastNode->next;
        }
        return temp->next;
    }   
};

File: leetcode_problems/merge_Two_sorted_lists.py
# Definition for singly-linked list.
from ast import List
from lib2to3.pgen2.token import OP
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        temp = ListNode(0, None)
        lastNode = temp
        while True:
            # If list1 is empty
            if list1 is None:
                lastNode.next = list2
                break
            # If list2 is empty
            elif list2 is None:
                lastNode.next = list1
                break
            
            if list1.val <= list2.val:
                lastNode.next = list1
                list1 = list1.next
            else:
                lastNode.next = list2
                list2 = list2.next

            lastNode = lastNode.next
        return temp.next


def printer(list: Optional[ListNode]):
    c = list
    while(c.next != None):
        print(c.val)
        c = c.next
    print(c.val)


def main():
    n3 = ListNode(2, None)
    n2 = ListNode(1, n3)
    n1 = ListNode(0, n2)
    m3 = ListNode(9, None)
    m2 = ListNode(8, m3)
    m1 = ListNode(7, m2)
    sol = Solution()
    sol.mergeTwoLists(m1, n1)


if __name__ == "__main__":
    main()
